gnovm/pkg/gnolang/walk.go

```diff
@@ -0,0 +1,422 @@
+			Walk(in, out, n.Init)
+		}
+		if n.X != nil {
+			Walk(in, out, n.X)
+		}
+		walkStmtList(in, out, n.bodyStmt.Body)
+
+	// case *TypeSwitchStmt:
+	//	if n.Init != nil {
+	//		Walk(in, n.Init)
+	//	}
+	//	Walk(in, n.Assign)
+	//	Walk(in, n.Body)
+
+	// case *CommClause:
+	//	if n.Comm != nil {
+	//		Walk(in, n.Comm)
+	//	}
+	//	walkStmtList(in, n.Body)
+
+	//case *SelectStmt:
+	//	Walk(in, n.Body)
+	//
+	case *ForStmt:
+		if n.Init != nil {
+			Walk(in, out, n.Init)
+		}
+		if n.Cond != nil {
+			Walk(in, out, n.Cond)
+		}
+		if n.Post != nil {
+			Walk(in, out, n.Post)
+		}
+		walkStmtList(in, out, n.Body)
+
+	case *RangeStmt:
+		if n.Key != nil {
+			Walk(in, out, n.Key)
+		}
+		if n.Value != nil {
+			Walk(in, out, n.Value)
+		}
+		Walk(in, out, n.X)
+		walkStmtList(in, out, n.Body)
+
+	// Declarations
+	case *ImportDecl:
+		if n.Label != "" {
+			Walk(in, out, Nx(n.Label))
+		}
+		if n.Name != "" {
+			Walk(in, out, &n.NameExpr)
+		}
+		//Walk(in, n.Path)
+		//if n.Comment != nil {
+		//	Walk(in, n.Comment)
+		//}
+
+	case *ValueDecl:
+		//if n.Doc != nil {
+		//	Walk(in, n.Doc)
+		//}
+
+		for i := range n.NameExprs {
+			Walk(in, out, &n.NameExprs[i])
+		}
+
+		if n.Type != nil {
+			Walk(in, out, n.Type)
+		}
+		walkExprList(in, out, n.Values)
+		//if n.Comment != nil {
+		//	Walk(in, n.Comment)
+		//}
+
+	case *TypeDecl:
+		//if n.Doc != nil {
+		//	Walk(in, n.Doc)
+		//}
+		Walk(in, out, &n.NameExpr)
+		//if n.TypeParams != nil {
+		//	Walk(in, n.TypeParams)
+		//}
+		Walk(in, out, n.Type)
+		//if n.Comment != nil {
+		//	Walk(in, n.Comment)
+		//}
+
+	//case *BadDecl:
+	//	// nothing to do
+	//
+	//case *GenDecl:
+	//	if n.Doc != nil {
+	//		Walk(in, n.Doc)
+	//	}
+	//	for _, s := range n.Specs {
+	//		Walk(in, s)
+	//	}
+	//
+	case *FuncDecl:
+		//if n.Doc != nil {
+		//	Walk(in, n.Doc)
+		//}
+		if n.Recv.Name != "" {
+			Walk(in, out, &n.Recv)
+		}
+		Walk(in, out, &n.NameExpr)
+		Walk(in, out, &n.Type)
+		if n.Body != nil {
+			walkStmtList(in, out, n.Body)
+		}
+	case *RefExpr:
+		Walk(in, out, n.X)
+
+	// Files and packages
+	case *FileNode:
+		//if n.Doc != nil {
+		//	Walk(in, n.Doc)
+		//}
+		Walk(in, out, Nx(n.Name))
+		walkDeclList(in, out, n.Decls)
+		// don't walk n.Comments - they have been
+		// visited already through the individual
+		// nodes
+
+	// case *Package:
+	//	for _, f := range n.Files {
+	//		Walk(in, f)
+	//	}
+
+	default:
+		panic(fmt.Sprintf("ast.Walk: unexpected node type %T", n))
+	}
+
+	if out != nil {
+		if out = out.Visit(node); out == nil {
+			return
+		}
+		out.Visit(nil)
+	}
+	if in != nil {
+		in.Visit(nil)
+	}
+}
+
+type inspector func(Node) bool
+
+func (f inspector) Visit(node Node) Visitor {
+	if f != nil && f(node) {
+		return f
+	}
+	return nil
+}
+
+// Inspect traverses an AST in depth-first order: It starts by calling
+// f(node); node must not be nil. If f returns true, Inspect invokes f
+// recursively for each of the non-nil children of node, followed by a
+// call of f(nil).
```

Comment:
This comment needs to be updated to reflect what `Inspect` really does now. The behaviour is now close to https://pkg.go.dev/golang.org/x/tools/go/ast/astutil#Apply, you could take inspiration from it.